Session Based Authentication

br41nslug opened the pull request:
## Scope

fixes https://github.com/directus/directus/issues/21118
fixes https://github.com/directus/directus/issues/21159

What's changed:

- Introduced a new authentication mode `"session"` which stores a session token as cookie.
- Adds `SESSION_COOKIE_*` environment variables, same subset as refresh_tokens.
- Implemented shares authentication in the SDK auth composable.
- Implemented the new auth mode in the auth middleware
- Updated SSO (open-id, saml, oath2) to default to the new "session" mode
- Updated the App to use the new authentication and removed obsolete token logic (using a regular `src` for `v-image` and removed the `addTokenToURL` util)
- Updated the SDK types to allow this new method

## TODO

- [x] Find missed docs pages to update

## Potential Risks / Drawbacks

- Authentication changes are always tricky, im sure i've missed an edge-case somewhere.

## Potential Advantages
- No more access_tokens in the URL for assets
- Credential sharing between multiple tabs of the app (or sdk) without one invalidating the refresh token of the other
- API calls done in the same browser as you're logged into the app will now work (but can be overwritten by an explicit Authorization header or access_token query param)

## Review Notes

- For SSO i am currently only able to test google's open-id myself!
- Considered adding CSRF but i dont think this is applicable here since we explicitly only accept `application/json` bodies on the API which cannot be done using regular forms (which is the main attack vector) and ajax calls are limited by cors/csp
- Also considered signing the cookie but since the JWT used as token is already signed and validated i dont't think thats explicitly required here
- I've excluded the WebSocket from using the cookie session as i am not sure I can currently secure that properly against cross site websockets

## Review Questions
- I've excluded the session cookie from the RFC6750 multiple auth methods error else this will break a lot of existing installs that are using the API from the same domain as the App. Do you agree or do we make this a breaking change?
- I've opted to let the SSO providers (openID, SAML, oauth2) default to the new "session" mode so they are compatible with the updates to the app, but of course this is a breaking change for anyone relying on the refresh token with SSO in an external app. For that i've added an env var. Is it right to prioritize making it work with the app here?

br41nslug committed:
extract cookie authentication

br41nslug committed:
use env vars

br41nslug committed:
added access cookie settings

br41nslug committed:
adjust app to not send double authentication

br41nslug committed:
updated cookie refreshing

br41nslug committed:
unified cookie env variables

br41nslug committed:
updated graphql login/refresh

br41nslug committed:
double checking clearing and setting of cookies and cleaning up obsolete type casting

br41nslug committed:
reverted env leftovers

br41nslug committed:
WIP refactoring

br41nslug committed:
switch app to session mode

br41nslug committed:
WIP auth refactor

br41nslug committed:
App auth using session

br41nslug committed:
updated sso providers

br41nslug committed:
updated sdk type

br41nslug committed:
cleaned up app token usage

br41nslug committed:
updated jwt util

br41nslug committed:
updated missing ldap validation

br41nslug committed:
extracted and reuse auth mode type

br41nslug committed:
updated docs

br41nslug committed:
small tweaks

br41nslug committed:
fixed using wrong token in the jwt

br41nslug committed:
updated default session ttl to 1 day

br41nslug committed:
updated some docs references

br41nslug committed:
Added breaking change placeholder

br41nslug committed:
making sso auth mode configurable

br41nslug committed:
cleaned up docs

br41nslug committed:
prettier

br41nslug committed:
Merge branch 'main' into session-token-authentication

br41nslug committed:
satisfy spellcheck

br41nslug committed:
satisfy spellcheck

br41nslug committed:
Merge branch 'main' into session-token-authentication

br41nslug committed:
Merge branch 'main' into session-token-authentication

br41nslug committed:
prettier

br41nslug committed:
Merge branch 'main' into session-token-authentication

br41nslug committed:
prettier

br41nslug committed:
Merge branch 'main' into session-token-authentication

azrikahar reviewed (commented):
> For SSO i am currently only able to test google's open-id myself!

Did a quick test with GitHub OAuth2 and wasn't able to spot any issues 👍

azrikahar commented on app/src/auth.ts:
Just noticed an error when manually logging out, we do send `mode: 'session'` in the body during login but seems like we're not sending it when logging out, so the API's `getCurrentMode` is interpreting mode as `cookie`. 

Should we post `mode: 'session'` during logout as well? Should be the line with `await api.post(`/auth/logout`);` if I'm not mistaken.

rijkvanzanten committed:
Update docs/extensions/app-composables.md

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

rijkvanzanten committed:
Update docs/reference/authentication.md

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
removed obsolete sentence part

br41nslug committed:
fixed number

br41nslug committed:
prettier

br41nslug committed:
Fixed logout mode in the docs, sdk, app, api and graphql

br41nslug committed:
small consistency fix

br41nslug committed:
Merge branch 'main' into session-token-authentication

paescuj commented on app/src/components/v-image.vue:
Not important, but `inheritAttrs: false` and `v-bind="attrs"` cancel each other out together.

FWIW: Since it's effectively just an alias for `img` now, we might as well remove that component all together, but better to do that in a separate PR 👍

paescuj commented on api/src/constants.ts:
This has already been used that way for `COOKIE_OPTIONS` before, but I think we should move the cookie options retrievement to an util like `getCookieOptions` where `useEnv()` is called inside that function instead of top-level to fetch the current state (relevant for later on where env can change during runtime).

`REFRESH_COOKIE_OPTIONS`  and `SESSION_COOKIE_OPTIONS` seems slightly misplaced in `constants.ts` anyway, seeing that all the other variables are plain constants (no function calls).

br41nslug replied:
Sounds good 👍 But lets do that in a separate PR

paescuj commented on app/src/api.ts:
```diff
@@ -1,9 +1,7 @@
 import axios, { AxiosError, AxiosResponse, InternalAxiosRequestConfig } from 'axios';
-import PQueue, { type Options, type QueueAddOptions } from 'p-queue';
-import { addQueryToPath } from './utils/add-query-to-path';
+import PQueue, { type DefaultAddOptions, type Options } from 'p-queue';
```
```suggestion
import PQueue, { type Options } from 'p-queue';
```

paescuj commented on docs/guides/extensions/interfaces-radio-selector-icons.md:
```diff
@@ -157,6 +157,12 @@ select-icon, which provides a searchable dropdown of the icon library.
 
+::: tip
+
+[DEPRECATED] Use session cookies instead.
```
Could we add a link? I guess to the SDK authentication guide

paescuj replied:
Oh right, to the API auth page 👍 ❤️

br41nslug replied:
Perhaps a link to the break change note is most appropriate 🤔 As whether this is enabled by default or not will depend on the version you're running

paescuj replied:
Great idea 🚀

paescuj commented on docs/guides/extensions/interfaces-radio-selector-icons.md:
```diff
@@ -157,6 +157,12 @@ select-icon, which provides a searchable dropdown of the icon library.
 
 ## Work With Images
 
+::: tip
+
+[DEPRECATED] Use session cookies instead.
```
```suggestion
::: warning DEPRECATED

Use session cookies instead.
```

like here https://docs.directus.io/reference/query.html#metadata

paescuj commented on docs/guides/extensions/modules-build-landing-page.md:
```diff
@@ -500,6 +500,12 @@ api.get(`/items/pages?fields=title,banner,content&filter[uri][_eq]=${page}`).the
 
 ### Work With Images
 
+::: tip
+
+[DEPRECATED] The query parameter approach used here is deprecated, make use of session cookies instead.
```
```suggestion
::: warning DEPRECATED

The query parameter approach used here is deprecated, make use of session cookies instead.
```

like here https://docs.directus.io/reference/query.html#metadata

paescuj replied:
Could we add a link? I guess to the SDK authentication guide

paescuj commented on docs/reference/authentication.md:
```diff
@@ -12,30 +12,46 @@ pageClass: page-reference
 
-Once you have your access token, there are two ways to pass it to the API, via the `access_token` query parameter, or in
-the request's Authorization Header.
+Once you have your access token, there are three ways to pass it to the API, in the request's Authorization Header, or
+as session cookie, or via the `access_token` query parameter.
```
```suggestion
Once you have your access token, there are three ways to pass it to the API: In the request's Authorization Header, as
session cookie, or via the `access_token` query parameter.
```

paescuj commented on docs/reference/authentication.md:
````diff
@@ -12,30 +12,46 @@ pageClass: page-reference
+?access_token=<token>
+```
+
+::: Tip
````
Might be worth a
```suggestion
::: warning
```

paescuj commented on docs/releases/breaking-changes.md:
```diff
@@ -13,6 +13,10 @@ these to a minimum, but rest assured we only make them with good reason.
 
 Starting with Directus 10.0, here is a list of potential breaking changes with remedial action you may need to take.
 
+## Version [NEXT]
```
```suggestion
## Version 10.10.0
```

paescuj commented on docs/self-hosted/config-options.md:
```diff
@@ -345,6 +345,11 @@ Redis is required when you run Directus load balanced across multiple containers
+| `SESSION_COOKIE_DOMAIN`             | Which domain to use for the session cookie. Useful for development mode.                                                                                                                             | --                        |
+| `SESSION_COOKIE_SECURE`             | Whether or not to use a secure cookie for the session token in cookie mode.                                                                                                                          | `false`                   |
+| `SESSION_COOKIE_SAME_SITE`          | Value for `sameSite` in the session token cookie when in cookie mode.                                                                                                                                | `lax`                     |
+| `SESSION_COOKIE_NAME`               | Name of session token cookie .                                                                                                                                                                       | `directus_session_token`  |
```
```suggestion
| `SESSION_COOKIE_NAME`               | Name of session token cookie.                                                                                                                                                                        | `directus_session_token`  |
```

br41nslug committed:
Update docs/guides/extensions/interfaces-radio-selector-icons.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

br41nslug committed:
Update docs/reference/authentication.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

br41nslug committed:
Update docs/self-hosted/config-options.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

br41nslug committed:
Update docs/guides/extensions/modules-build-landing-page.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

br41nslug committed:
Update app/src/api.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

paescuj commented:
> * I've excluded the session cookie from the RFC6750 multiple auth methods error else this will break a lot of existing installs that are using the API from the same domain as the App. Do you agree or do we make this a breaking change?

I agree, I don't see this as a strict requirement. Seems to be a good interim solution 👍 

> * I've opted to let the SSO providers (openID, SAML, oauth2) default to the new "session" mode so they are compatible with the updates to the app, but of course this is a breaking change for anyone relying on the refresh token with SSO in an external app. For that i've added an env var. Is it right to prioritize making it work with the app here?

Yes, I think this is the right way! Makes sense to adjust this at this point in time.

> * I didnt want to add an breaking changes all over the place so for the `AuthenticationService` `login` and `refresh` function signatures but I do think changing the `otp` and `session` properties to a shared config object would be nicer if we agree on that breaking change.

I think it's acceptable to add this to the list of breaking changes - better to do so now, IMO. If I understand correctly, this seems to make it more future proof 👍

br41nslug committed:
Removed obsolete v-image code

br41nslug committed:
updated note

br41nslug committed:
updated note

br41nslug committed:
text change

br41nslug committed:
Updated authentication service signatures

br41nslug committed:
updated authentication service usages

br41nslug committed:
updated breaking change

br41nslug committed:
Migrated conflicting commits from pascal

br41nslug committed:
prettier

br41nslug committed:
satisfy spellchecker

br41nslug committed:
Create odd-buses-nail.md

paescuj commented on docs/releases/breaking-changes.md:
```diff
@@ -13,6 +13,81 @@ these to a minimum, but rest assured we only make them with good reason.
+this over the previous token based authentication. This impacts `oauth2`, `open-id` and `saml` SSO installations as they
+too will now also default to the new session based authentication in order to work with the App out-of-the-box.
+
+To keep using the old SSO functionality setting the refresh token instead of session token you can set
```
We need to mention that this would only work for external apps and not when used for the Data Studio

paescuj replied:
If there should be a need for both methods, I think it might be possible to set-up the same provider two times under a different name 🤔

paescuj committed:
Apply same payload structure in auth.ts

paescuj committed:
Small adjustments to breaking changes entries

paescuj commented on docs/releases/breaking-changes.md:
```diff
@@ -13,6 +13,83 @@ these to a minimum, but rest assured we only make them with good reason.
+this method over the previous token based authentication. This impacts `oauth2`, `open-id` and `saml` SSO installations
+as they too will now default to the new session based authentication in order to work with the App out-of-the-box.
+
+To keep using the old SSO functionality setting the refresh token instead of session token, you can set
+`AUTH_<PROVIDER>_MODE=cookie`.
```
I think we need to mention that this would only work for external apps and not when used for the Data Studio.

If there should be a need for both methods (Data Studio access & external app with cookie mode), theoretically it should be possible to set-up the same provider two times under a different name. Although, I'm afraid there's currently no option to hide a provider in the Data Studio.

paescuj commented on .changeset/odd-buses-nail.md:
```diff
@@ -0,0 +1,7 @@
+---
+"@directus/app": major
+"@directus/api": major
+"@directus/sdk": patch
```
```suggestion
"@directus/sdk": patch
"@directus/env": patch
```

paescuj committed:
Include `@directus/env` in changeset

paescuj commented on docs/releases/breaking-changes.md:
```diff
@@ -13,6 +13,83 @@ these to a minimum, but rest assured we only make them with good reason.
+
+### Session Cookie Based Authentication
+
+For improved security and ease of use we have implemented session based authentication and have updated the App to use
```
We should probably also mention the new `SESSION_COOKIE_*` configs

paescuj reviewed (approved):
Besides the two docs comments / change requests -> LGTM 🥳

Tests:
- [x] Auth drivers (with session / cookie mode)
  - [x] Local (obviously)
  - [x] LDAP (test server)
  - [x] OpenID (KeyCloak)
  - [x] OAuth2 (GitHub)
  - [x] SAML (test server from blackbox tests)
- [x] GraphQL auth mutations
- [x] Data & Dashboard exports
- [x] Shares
- [x] Images / Files
  - [x] File Library (File Editor, and co.)
  - [x] Block Editor
  - [x] WYSIWYG (with and without static token)
  - [x] Image Interfaces / Displays
- [x] Refresh (via low TTL)
- [x] Docs checked

rijkvanzanten committed:
Add breaking change note for content versioning

rijkvanzanten committed:
nothing to see here

br41nslug committed:
updated change note

br41nslug committed:
updated change note

paescuj committed:
Merge branch 'main' into session-token-authentication

paescuj committed:
Adjust heading levels in breaking changes

paescuj committed:
Final fixes & clean-up

- Finalize docs
- Update env-stub
- Clarify missing refresh token error message (to not give the impression that session token can be sent in the payload)
- Fix & align share refresh
- Clarify / clean-up some comments
- Enhance types
- Fix / finalize / align GraphQL auth mutations
- Update specs
- Update SDK auth method comments
- Adjust blackbox tests
